internal/resource/url.go

```diff
@@ -278,6 +288,83 @@ func (f *Fetcher) FetchFromOEM(u url.URL, dest *os.File) ([]byte, error) {
 	return f.copyAndHash(dest, src)
 }
 
+// FetchFromS3 gets data from an S3 bucket as described by u and writes it into
+// dest, returning a hash (if one is set) or an error. It will attempt to
+// acquire IAM credentials from the EC2 metadata service, and if this fails will
+// attempt to fetch the object with anonymous credentials.
+func (f *Fetcher) FetchFromS3(u url.URL, dest *os.File) ([]byte, error) {
+	ctx := context.Background()
+	if f.Client != nil && f.Client.timeout != 0 {
+		var cancelFn context.CancelFunc
+		ctx, cancelFn = context.WithTimeout(ctx, f.Client.timeout)
+		defer cancelFn()
+	}
+
+	// Determine the region this bucket is in
+	sess, err := session.NewSession()
+	if err != nil {
+		return nil, err
+	}
+	region, err := s3manager.GetBucketRegion(ctx, sess, u.Host, "us-east-1")
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok && aerr.Code() == "NotFound" {
+			return nil, fmt.Errorf("couldn't determine the region for bucket %q: %v", u.Host, err)
+		}
+		return nil, err
+	}
+
+	// Create a new session with the discovered region
+	sess, err = session.NewSession(&aws.Config{
+		Region: aws.String(region),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	creds := credentials.AnonymousCredentials
+	if f.OEM == "ec2" {
```

Comment:
Where is `f.OEM` assigned? Is there any way we can avoid having to plumb the OEM everywhere?